p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,344 @@
+			if err != nil {
+				log.Warnf("could not detach datachannel: id: %d", *dc.ID())
+				return
+			}
+			w := msgio.NewWriter(rwc)
+			select {
+			case conn.acceptQueue <- acceptStream{rwc, dc}:
+			default:
+				log.Warnf("connection busy, rejecting stream")
+				// reject stream without instantiating a delimited writer
+				b, _ := proto.Marshal(&pb.Message{Flag: pb.Message_RESET.Enum()})
+				w.WriteMsg(b)
+				rwc.Close()
+			}
+		})
+	})
+
+	return conn, nil
+}
+
+func (c *connection) resetStreams() {
+	if c.IsClosed() {
+		return
+	}
+	c.m.Lock()
+	defer c.m.Unlock()
+	for k, stream := range c.streams {
+		// reset the streams, but we do not need to be notified
+		// of stream closure
+		stream.close(true, false)
+		delete(c.streams, k)
+	}
+
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{
+		Transport: "webrtc",
+	}
+}
+
+// Close closes the underlying peerconnection.
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	return c.pc.Close()
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *connection) OpenStream(ctx context.Context) (network.MuxedStream, error) {
+	if c.IsClosed() {
+		return nil, os.ErrClosed
+	}
+
+	streamID, err := c.idAllocator.nextID()
+	if err != nil {
+		return nil, err
+	}
+	dc, err := c.pc.CreateDataChannel("", &webrtc.DataChannelInit{ID: streamID})
+	if err != nil {
+		return nil, err
+	}
+	rwc, err := c.detachChannel(ctx, dc)
+	if err != nil {
+		return nil, fmt.Errorf("open stream: %w", err)
+	}
+	stream := newStream(
+		c,
+		dc,
+		rwc,
+		nil,
+		nil,
+	)
+	c.addStream(stream)
+	return stream, nil
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case accStream := <-c.acceptQueue:
+		stream := newStream(
+			c,
+			accStream.channel,
+			accStream.stream,
+			nil,
+			nil,
+		)
+		c.addStream(stream)
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+func (c *connection) LocalPrivateKey() ic.PrivKey {
+	return c.privKey
+}
+
+func (c *connection) RemotePeer() peer.ID {
+	return c.remotePeer
+}
+
+func (c *connection) RemotePublicKey() ic.PubKey {
+	return c.remoteKey
+}
+
+// implement network.ConnMultiaddrs
+func (c *connection) LocalMultiaddr() ma.Multiaddr {
+	return c.localMultiaddr
+}
+
+func (c *connection) RemoteMultiaddr() ma.Multiaddr {
+	return c.remoteMultiaddr
+}
+
+// implement network.ConnScoper
+func (c *connection) Scope() network.ConnScope {
+	return c.scope
+}
+
+func (c *connection) Transport() tpt.Transport {
+	return c.transport
+}
+
+func (c *connection) addStream(stream *webRTCStream) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.streams[stream.id] = stream
```

Comment:
Hi @marten-seemann it means I saw your comment and will address it. I was out due to sickness, but I'm getting back on the saddle as we speak, so will address your issue this week!

Reply:
This is a private function called from the same object only when a 1) new stream is initiated by the remote, or 2) we create a new stream. In case 1 the stream ID is extracted from the incoming datachannel, so it is guaranteed to exist. In case 2, we get the id from the `sidAllocator` which already performs the odd even check, and `addStream` is only called after a datachannel is created with the new stream ID. 

Finally, since this function is called for both outgoing and incoming streams, we need not perform an odd-even check here since it  has been done by the stream ID allocator. If such a check must be performed (and it should), it should be done when the remote creates a new datachannel.

In both cases, the stream ID is guaranteed to exist.

Reply:
@ckousik This is not correct. You can't just assume that the peer is behaving correctly. A malicious peer might send us datachannels with IDs that don't conform to the even / odd rule, or IDs that already exist. Here's a patch that will make the client create datachannel 5, and the server happily accepting it (can be used with my [Gist](https://gist.github.com/marten-seemann/1b2507d984b9cf637475c995081094ec)):
```diff
diff --git a/p2p/transport/webrtc/connection.go b/p2p/transport/webrtc/connection.go
index c5928e6a..13955ba3 100644
--- a/p2p/transport/webrtc/connection.go
+++ b/p2p/transport/webrtc/connection.go
@@ -100,6 +100,11 @@ func newConnection(
                        return
                }
                dc.OnOpen(func() {
+                       pers := "server"
+                       if direction == network.DirOutbound {
+                               pers = "client"
+                       }
+                       fmt.Printf("%s OnDataChannel: %d\n", pers, *(dc.ID()))
                        rwc, err := dc.Detach()
                        if err != nil {
                                log.Warnf("could not detach datachannel: id: %d", *dc.ID())
@@ -343,26 +348,30 @@ func (c *connection) setRemotePublicKey(key ic.PubKey) {
 // By definition, the DTLS role for inbound connections is set to DTLS Server,
 // and outbound connections are DTLS Client.
 type sidAllocator struct {
-       n atomic.Uint32
+       dir network.Direction
+       n   atomic.Int32
 }
 
 func newSidAllocator(direction network.Direction) *sidAllocator {
+       a := new(sidAllocator)
        switch direction {
        case network.DirInbound:
                // server will use odd values
-               a := new(sidAllocator)
                a.n.Store(1)
-               return a
        case network.DirOutbound:
                // client will use even values
-               return new(sidAllocator)
        default:
                panic(fmt.Sprintf("create SID allocator for direction: %s", direction))
        }
+       a.dir = direction
+       return a
 }
 
 func (a *sidAllocator) nextID() (*uint16, error) {
        nxt := a.n.Add(2)
+       if a.dir == network.DirOutbound && nxt > 2 { // don't mess with the handshake stream
+               nxt = a.n.Add(1)
+       }
        if nxt > math.MaxUint16 {
                return nil, fmt.Errorf("sid exhausted")
        }

```

Reply:
This assumption has been removed.